pkgs/development/haskell-modules/configuration-nix.nix

```diff
@@ -1329,4 +1329,23 @@ self: super: builtins.intersectAttrs super {
       gi-webkit2
       gi-webkit2webextension
       ;
+
+  # Default MPI flags force MPICH linking but OpenMPI runtime, thus breaking the package.
+  # Override forces usage of correct mpi attribute, -fopenmpi-debian generically links against -lmpi
+  inherit (lib.mapAttrs (_: overrideCabal {
+      librarySystemDepends = [ pkgs.mpi ];
+      executableSystemDepends = [ pkgs.mpi ];
+      testSystemDepends = with pkgs; [ mpi openssh mpiCheckPhaseHook ];
+      extraLibraries = [ pkgs.mpi ];
+      configureFlags = [
+        "--extra-include-dirs=${pkgs.mpi}/include"
+        "--extra-lib-dirs=${pkgs.mpi}/lib"
+        "-fopenmpi-debian" # This links against "-lmpi" and also works for MPICH
+      ];
```

Comment:
It seems like `mpi` has pkgconfig files available.

Maybe you could send a PR upstream making the .cabal files use the pkgconfig files?  That would be easier than all these overrides.  It is definitely a code smell that upstream hard-codes links to `/usr` in their `.cabal` file.

--------------------------------------------

It is strange having to override all of `librarySystemDepends`, `executableSystemDepends`, `testSystemDepends`, `extraLibraries`, and `configureFlags`.  I would have expected you to only need to override something like the `buildDepends` argument.

Also, you'll probably want to remove the `mpich` argument from the generated derivation (and possibly replace it with `mpi`?), which you can do in a PR to `cabal2nix`.

Reply:
Oh, and one other thing: we generally ask for Haskell-related PRs to be made to the `haskell-updates` branch.  Could you rebase this PR on `haskell-updates`?

Reply:
> Maybe you could send a PR upstream making the .cabal files use the pkgconfig files? 

I can try tomorrow and will link. But the upstream project seems a little inactive, last commit is from beginning 2020, so let's see :)

`configureFlags` in the current setup is required to get the `-fopenmpi-debian` flag, which actually allow to link generically against `-lmpi` working for all major MPI implementations.

Wrt to the other overrides it is very well possible that I'm not understanding the haskell setup in nixpkgs well enough and maybe you can give me some hints there and comment on them? :slightly_smiling_face: 

- I've added mpi it to `executableSystemDepends` as the `mpiexec` executable will be required at runtime for every programme and library using `mpi-hs` packages
- I think `mpi` is not passed properly to the generated derivation. As far as I understand, `cabal2nix` is called first and determines which packages are passed to the nix expression building the actual haskell package, correct? I think at this stage `mpich` is still passed, because this is the default flag in the cabal file. Thus, when I override the `configureFlags` later to use `-fopenmpi-debian`, which switches to `mpi` instead of the `mpihs` dependency, `mpich` is already passed because it happened a stage earlier. Do I understand this correctly? :grimacing: 
- Is it enough to just pass `pkgs.mpi` to `buildDepends`? During building `-lmpi` is required, for compilation probably the `mpicc` compiler wrapper and during the test phase **and** runtime `mpiexec`.

Reply:
> Also, you'll probably want to remove the mpich argument from the generated derivation (and possibly replace it with mpi?), which you can do in a PR to cabal2nix. (Feel free to ask if you need pointers for this)

I need a pointer for this :smile: I assume it means this influences which attributes are passed to `mkDerivation` and I can get rid of `mpich` in first place?

Reply:
> Maybe you could send a PR upstream making the .cabal files use the pkgconfig files? That would be easier than all these overrides. It is definitely a code smell that upstream hard-codes links to /usr in their .cabal file.

https://github.com/eschnett/mpi-hs/pull/3 :slightly_smiling_face:

Reply:
While writing the PR upstream I've tried to nixify the mpi-hs repo directly and the best solution I've could come up with so far was something like this:

```nix
{ mkDerivation
, pkg-config
, base
, bytestring
, c2hs
, lib
, monad-loops
, mpi
, mpiCheckPhaseHook
, openssh
}:
mkDerivation {
  pname = "mpi-hs";
  version = "0.7.2.0";
  src = lib.cleanSource ./..;
  isLibrary = true;
  isExecutable = true;
  buildTools = [ pkg-config ];
  pkg-configDepends = [ mpi ];
  libraryHaskellDepends = [ base bytestring monad-loops ];
  libraryToolDepends = [ c2hs ];
  executableHaskellDepends = [ base ];
  executableSystemDepends = [ mpi ];
  testHaskellDepends = [ base monad-loops ];
  testSystemDepends = [ openssh mpi mpiCheckPhaseHook] ;
  homepage = "https://github.com/eschnett/mpi-hs#readme";
  description = "MPI bindings for Haskell";
  license = lib.licenses.asl20;
}
```
(manually edited result of `cabal2nix`. But currently I've no idea how to do something like this in the nixpkgs infrastructure :grimacing:

Reply:
> Oh, and one other thing: we generally ask for Haskell-related PRs to be made to the haskell-updates branch. Could you rebase this PR on haskell-updates?

Done :+1:

Reply:
With the upstream PR merged, do we need a new release on hackage or do we pull in the commits as patches?

Reply:
The easiest thing for us would be to get a new release on Hackage.

Maybe you could ask to be a co-maintainer of the package?  That way you wouldn't have to ask the current maintainer for doing things like bumping versions, releasing to hackage, etc.

If the maintainer doesn't want to do a release to hackage, then the alternative is to apply patches in Nixpkgs (but we try to avoid this when possible).

Reply:
Alright, I'm second maintainer now and we have a new version on Hackage https://hackage.haskell.org/package/mpi-hs-0.7.3.0

Do we need to wait for the next Hackage update?

Reply:
Nice, glad you got a new release out!

> Do we need to wait for the next Hackage update?

Yeah, the easiest thing to do is to wait for the next Hackage update, and then send a PR marking `mpi-hs` as unbroken (and possibly adding any overrides that are still necessary).

It should be the PR after https://github.com/NixOS/nixpkgs/pull/255037.

Reply:
Alright, so mpi-hs 0.7.3.0 has landed in #259060, although as `mpi-hs_0_7_3_0`. I was thinking `mpi-hs` should be somewhat like 

```nix
let addMpiHsTestDeps = drv: addTestToolDepends
      [ pkgs.openssh pkgs.mpiCheckPhaseHook ]
      drv;
in mpi-hs = addMpiHsTestDeps (
          if pkgs.mpi.pname == "openmpi"
          then callCabal2nixWithOptions "mpi-hs" mpi-hs_src "-fopenmpi -f-mpich -f-mvapich" { ompi = pkgs.mpi; }
          else if pkgs.mpi.pname == "mpich"
          then callCabal2nixWithOptions "mpi-hs" mpi-hs_src "-f-openmpi -fmpich -f-mvapich" { mpich = pkgs.mpi; }
          else if pkgs.mpi.pname == "mvapich2"
          then callCabal2nixWithOptions "mpi-hs" mpi-hs_src "-f-openmpi -f-mpich -fmvapich" { mvapich2 = pkgs.mpi; }
          else callCabal2nixWithOptions "mpi-hs" mpi-hs_src "-f-openmpi -f-mpich -f-mvapich" { mpi = pkgs.mpi; }
        );
```
However, I think I don't really understand how to call a package in this infrastructure with options. I couldn't find another place where `callCabal2nixWithOptions` is used in the haskell infrastructure within nixpkgs. Also what to do with mpi-hs vs mpi-hs_0_7_3_0? :)

Reply:
callCabal2nix(withOptions) uses "import from derivation" and is therefore not allowed in nixpkgs.
We generally modify packages with the helpers listed in the nixpkgs manual first and foremost `overrideCabal`.

Reply:
Alright, understandable. I came up with a different solution (and pushed it), but have to admit it feels somewhat hacky. As the expression is generated from the default flags, formally `mpi-hs` always has `ompi` (OpenMPI) as pkg-config dependency. However, the purpose here is exactly to make the tests pass and make it work with the other MPI implementations. Thus, I'm now changing cabal flags via `enableCabalFlag` and `disableCabalFlag` and override `ompi` to always use `pkgs.mpi`. However, the pkg-config dependency should of course correctly be `mpich` or `mvapich` in the other cases. But this seems a formality. Please have another look and let me know what you think :)